internal/cloud/openstack/openstack.go

```diff
@@ -0,0 +1,246 @@
+			UserDomainName: userDomainName,
+			Username:       username,
+			Password:       password,
+		},
+	}
+
+	serversClient, err := clientconfig.NewServiceClient("compute", clientOpts)
+	if err != nil {
+		return nil, fmt.Errorf("creating compute client: %w", err)
+	}
+	serversClient.Microversion = microversion
+
+	subnetsClient, err := clientconfig.NewServiceClient("network", clientOpts)
+	if err != nil {
+		return nil, fmt.Errorf("creating network client: %w", err)
+	}
+	subnetsClient.Microversion = microversion
+
+	return &Cloud{
+		imds: imds,
+		api: &apiClient{
+			servers: serversClient,
+			subnets: subnetsClient,
+		},
+	}, nil
+}
+
+// Self returns the metadata of the current instance.
+func (c *Cloud) Self(ctx context.Context) (metadata.InstanceMetadata, error) {
+	name, err := c.imds.name(ctx)
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("getting name: %w", err)
+	}
+	providerID, err := c.imds.providerID(ctx)
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("getting provider id: %w", err)
+	}
+	role, err := c.imds.role(ctx)
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("getting role: %w", err)
+	}
+	vpcIP, err := c.imds.vpcIP(ctx)
+	if err != nil {
+		return metadata.InstanceMetadata{}, fmt.Errorf("getting vpc ip: %w", err)
+	}
+
+	return metadata.InstanceMetadata{
+		Name:       name,
+		ProviderID: providerID,
+		Role:       role,
+		VPCIP:      vpcIP,
+	}, nil
+}
+
+// List returns the metadata of all instances belonging to the same Constellation cluster.
+func (c *Cloud) List(ctx context.Context) ([]metadata.InstanceMetadata, error) {
+	uid, err := c.imds.uid(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("getting uid: %w", err)
+	}
+
+	uidTag := fmt.Sprintf("constellation-uid-%s", uid)
+
+	subnet, err := c.getSubnetCIDR(uidTag)
+	if err != nil {
+		return nil, err
+	}
+
+	srvs, err := c.getServers(uidTag)
+	if err != nil {
+		return nil, err
+	}
+
+	var result []metadata.InstanceMetadata
+	for _, s := range srvs {
+		if s.Name == "" {
+			continue
+		}
+		if s.ID == "" {
+			continue
+		}
+		if s.Tags == nil {
+			continue
+		}
+
+		var serverRole role.Role
+		for _, t := range *s.Tags {
+			if strings.HasPrefix(t, "constellation-role-") {
+				serverRole = role.FromString(strings.TrimPrefix(t, "constellation-role-"))
+				break
+			}
+		}
+		if serverRole == role.Unknown {
+			continue
+		}
+
+		subnetAddrs, err := parseSeverAddresses(s.Addresses)
+		if err != nil {
+			return nil, fmt.Errorf("parsing server %q addresses: %w", s.Name, err)
+		}
+
+		var vpcIP string
+		// In a best effort approach, we take the first fixed IPv4 address that is in the subnet
+		// belonging to our cluster.
+		for _, serverSubnet := range subnetAddrs {
+			for _, addr := range serverSubnet.Addresses {
+				if addr.Type != fixedIP {
+					continue
+				}
+
+				if addr.IPVersion != ipV4 {
+					continue
+				}
+
+				if addr.IP == "" {
+					continue
+				}
```

Comment:
I would write this as one If statement, can also be kept as is if you prefer it this way

Reply:
I prefer it this way, much faster to read imo.